qiskit_ibm_runtime/runtime_job.py

```diff
@@ -308,24 +309,48 @@ def _set_status(self, job_response: Dict) -> None:
             IBMError: If an unknown status is returned from the server.
         """
         try:
-            self._status = API_TO_JOB_STATUS[job_response["status"].upper()]
+            if "reason" in job_response["state"]:
+                self._reason = job_response["state"]["reason"]
+            status = job_response["state"]["status"]
+            self._state["status"] = self._get_api_msg(
+                status, API_TO_JOB_STATUS, self._reason
+            )
         except KeyError:
-            raise IBMError(f"Unknown status: {job_response['status']}")
+            raise IBMError(f"Unknown status: {status}")
 
     def _set_error_message(self, job_response: Dict) -> None:
         """Set error message if the job failed.
 
         Args:
             job_response: Job response from runtime API.
         """
-        if self._status == JobStatus.ERROR:
+        if self._state["status"] == JobStatus.ERROR:
             job_result_raw = self._api_client.job_results(job_id=self.job_id)
-            self._error_message = API_TO_JOB_ERROR_MESSAGE[
-                job_response["status"].upper()
-            ].format(self.job_id, job_result_raw)
+            self._error_message = self._get_api_msg(
+                job_response["state"]["status"], API_TO_JOB_ERROR_MESSAGE, self._reason
+            ).format(self.job_id, job_result_raw)
         else:
             self._error_message = None
 
+    def _get_api_msg(
+        self,
+        status: str,
+        msg_type: Dict,
+        reason: str = None,
+    ) -> str:
+        """Return the job status or error message.
+
+        Args:
+            status: Job status returned from the API.
+            msg_type: Dictionary to use to convert API message.
+            reason: Job status reason returned from the API.
+
+        Returns: A job status or errror message.
+        """
+        if reason:
+            status = status + " - " + reason
+        return msg_type[status.upper()]
```

Comment:
I believe this value should not return a concatenated string but one of the [JobStatus enum values](https://github.com/Qiskit/qiskit-terra/blob/2eee56616d50a9e26756f855ef4aa0135920ad78/qiskit/providers/jobstatus.py#L18-L27)

Reply:
Here's pseudo code that outlines the key changes that I believe are required:

```python
# change 1 - update mapping table that relies on the old concatenated status property
API_TO_JOB_STATUS = {
    "QUEUED": JobStatus.QUEUED,
    "RUNNING": JobStatus.RUNNING,
    "COMPLETED": JobStatus.DONE,
    "FAILED": JobStatus.ERROR,
    "CANCELLED": JobStatus.CANCELLED,
    # removed key "CANCELLED - RAN TOO LONG"
}

# change 2 - add dedicated function to derive status enum value from job response
self._status = status_from_job_response(response)


# change 3 - implement mapping code
def status_from_job_response(response):
  mapped_job_status = API_TO_JOB_ERROR_MESSAGE[response.state.status]
  if mapped_job_status == JobStatus.CANCELLED and response.reason = "Ran too long"
     mapped_job_status = JobStatus.ERROR
```

Reply:
How would this work if there are different reasons for different statuses in the future? Wouldn't we have to add more if statements?

Reply:
Yes, if we stick with the existing pattern to map the response from the API to a Qiskit `JobStatus` enum, we have to maintain some sort of mapping code. Ideally this mapping does not depend on a reason string but I don't know if that's feasible. My proposal was meant to map the existing code in a backward-compatible way.

Reply:
I made some updates - I think getting the [error message](https://github.com/Qiskit/qiskit-ibm-runtime/blob/main/qiskit_ibm_runtime/constants.py#L29) from the job response will still require the reason string unless we also want to refactor that logic